### PR TITLE
HHH-18951 Support custom annotation attributes as configuration parameters of Generator

### PR DESCRIPTION
**Pull request by quaff**
----------------------
By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt)
and can be relicensed under the terms of the [LGPL v2.1 license](https://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt) in the future at the maintainers' discretion.
For more information on licensing, please check [here](https://github.com/hibernate/hibernate-orm/blob/main/CONTRIBUTING.md#legal).

----------------------

https://hibernate.atlassian.net/browse/HHH-18951

### CODE REVIEW

**Comment by quaff** [comment]
@gavinking Please review.

File: hibernate-core/src/main/java/org/hibernate/internal/util/AnnotationUtil.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+	private static final List<String> excludedMethodNames = List.of( "hashCode", "toString", "annotationType" );
+
+	/**
+	 * Retrieve the given annotation's attributes as a {@link Map}, preserving all
+	 * attribute types.
+	 * @param annotation the annotation to retrieve the attributes for
+	 * @return the Map of annotation attributes, with attribute names as keys and
+	 * corresponding attribute values as values (never {@code null})
+	 */
+	public static Map<String, Object> getAttributes( Annotation annotation ) {
```
Comment (gavinking) [inline hibernate-core/src/main/java/org/hibernate/internal/util/AnnotationUtil.java:31]: `Map<String,Object>` is what we've just spent years trying to get away from.

**Comment by gavinking** [comment]
So I think this patch is in some sense perfectly fine and reasonable, but the thing is that `Configurable`, like the rest of the package it belongs to, is old stuff that we've already moved away from (except for the built-in generators). I don't think we should be giving it a second lease on life.

We moved to the new approach in order to achieve type safety, and this change undermines that.

So I guess I'm not really very keen on this.

**Comment by quaff** [comment]
> We moved to the new approach in order to achieve type safety, and this change undermines that.

How about introduce a new functional interface accept the annotation before configure?

**Comment by beikov** [comment]
Please create PRs primarily against the `main` branch, especially if it's an improvement or new feature.

**Comment by quaff** [comment]
> Please create PRs primarily against the `main` branch, especially if it's an improvement or new feature.

OK, should I create another PR or modify this one? I'd like it's fixed in 6.6.

**Comment by gavinking** [comment]
> I'd like it's fixed in 6.6.

That's _definitely_ not going to happen. 6.6 is finished and done.

But I'm not even convinced we should do this at all.

**Comment by quaff** [comment]
Alternatively, we could put the annotation self into parameters instead of its attributes.

**Comment by gavinking** [comment]
> Alternatively, we could put the annotation self into parameters instead of its attributes.

That's certainly *better*.

**Comment by gavinking** [comment]
Well, 

> > Alternatively, we could put the annotation self into parameters instead of its attributes.
> 
> That's certainly _better_.

Well, I dunno ... maybe not. It's somewhat more typesafe, at least.

**Comment by gavinking** [comment]
> How about introduce a new functional interface accept the annotation before configure?

I think if we were going to do this, this would be the way, because the interface would be parameterized by the annotation type, and it would be type safe. But it means adding an additional interface and additional complexity for something that I'm just not convinced we have a great justification for.

**Comment by gavinking** [comment]
> > How about introduce a new functional interface accept the annotation before configure?
> 
> I think if we were going to do this, this would be the way, because the interface would be parameterized by the annotation type, and it would be type safe. But it means adding an additional interface and additional complexity for something that I'm just not convinced we have a great justification for.

Actually @quaff, I've been forgetting something. We _already have an interface like this_, and from looking at the code its `initialize()` method _already gets called_ after a `Generator` is instantiated via the `BeanContainer`.

So I believe the functionality you're asking for already exists (and is even documented), I had just forgotten about it because I'm swapped-out on this stuff.

**Comment by quaff** [comment]
> > > How about introduce a new functional interface accept the annotation before configure?
> > 
> > 
> > I think if we were going to do this, this would be the way, because the interface would be parameterized by the annotation type, and it would be type safe. But it means adding an additional interface and additional complexity for something that I'm just not convinced we have a great justification for.
> 
> Actually @quaff, I've been forgetting something. We _already have an interface like this_, and from looking at the code its `initialize()` method _already gets called_ after a `Generator` is instantiated via the `BeanContainer`.
> 
> So I believe the functionality you're asking for already exists (and is even documented), I had just forgotten about it because I'm swapped-out on this stuff.

My bad, you are right.

**Comment by quaff** [comment]
> and is even documented

@gavinking The document mentioned `AnnotationBasedGenerator`, it should elaborate for such user case.

**Pull request closed without merging**